braindecode/preprocessing/eegprep_preprocess.py

```diff
@@ -0,0 +1,1139 @@
+"""Preprocessors using the EEGPrep package."""
+
+# Authors: Christian Kothe <christian.kothe@intheon.io>
+#
+# License: BSD-3
+
+import base64
+import json
+import logging
+import re
+from types import ModuleType
+from typing import Any, Sequence
+
+import numpy as np
+from mne import BaseEpochs
+from mne.io import BaseRaw
+
+from braindecode.preprocessing.preprocess import Preprocessor
+
+log = logging.getLogger(__name__)
+
+__all__ = [
+    "EEGPrep",
+    "RemoveDCOffset",
+    "Resampling",
+    "RemoveFlatChannels",
+    "RemoveDrifts",
+    "RemoveBadChannels",
+    "RemoveBadChannelsNoLocs",
+    "RemoveBursts",
+    "RemoveBadWindows",
+    "ReinterpolateRemovedChannels",
+    "RemoveCommonAverageReference",
+]
+
+try:
+    import eegprep
+except ImportError:
+
+    class EEGPrepMissing(ModuleType):
+        def __getattr__(self, _: str) -> Any:
+            raise ImportError(
+                "The eegprep package is required to use the EEGPrep preprocessor.\n"
+                "  Please install braindecode with the [eegprep] extra added as in\n"
+                "  'pip install braindecode[eegprep]' to use this functionality,\n"
+                "  or run 'pip install eegprep[eeglabio]' directly."
+            )
+
+    eegprep = EEGPrepMissing("eegprep")
+
+
+# Use a unique marker for embedding structured data in info['description']
+_MARKER_PATTERN = re.compile(r"<!-- eegprep-data:\s*(\S+)\s*-->", re.DOTALL)
+_MARKER_START = "<!-- eegprep-data:"
+_MARKER_END = "-->"
+# Marker key for empty numpy arrays
+_EMPTY_ARRAY_KEY = "__empty_array__"
+
+
+def _numpy_empty_decoder(dct):
+    """JSON decoder that reconstructs empty numpy arrays."""
+    if dct.get(_EMPTY_ARRAY_KEY):
+        return np.array([])
+    return dct
+
+
+def _encode_payload(data: dict) -> str:
+    """Serializes, encodes, and formats data into a marker string."""
+
+    class NumpyEmptyEncoder(json.JSONEncoder):
+        """JSON encoder that handles empty numpy arrays
+        (can occur as a sentinel object in the payload)."""
+
+        def default(self, obj):
+            # note we don't expect other numpy objects here besides np.array([])
+            if isinstance(obj, np.ndarray) and obj.size == 0:
+                return {_EMPTY_ARRAY_KEY: True}
+            return super().default(obj)
+
+    json_str = json.dumps(data, cls=NumpyEmptyEncoder)
+    encoded = base64.b64encode(json_str.encode("utf-8")).decode("ascii")
+    return f"{_MARKER_START} {encoded} {_MARKER_END}"
```

Comment:
It is strange this for me, an function that define class that define a function. Can you deconvolute and flat this?